### PR TITLE
feat: add comparison operators

### LINKED ISSUE #112: Support for comparison of bignums
### Problem

I want to compare 2 bignums

### Happy Case

I can compare 2 bignums. Implement `std::cmp::Ord` for `BigNum`

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Project Impact

None

### Blocker Context

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by kashbrti**
# Description
resolves #112 

## Problem\*

Resolves 

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: src/fns/constrained_ops.nr [hunk 1]
```diff
@@ -331,27 +332,11 @@ pub(crate) fn validate_gt<let N: u32, let MOD_BITS: u32>(lhs: [Field; N], rhs: [
     // so we do... p - x - r = 0 and there might be borrow flags
     // a - b = r
     // p + a - b - r = 0
-    let (result, carry_flags, borrow_flags) = unsafe { __validate_gt_remainder(lhs, rhs) };
+    let (underflow, result, carry_flags, borrow_flags) =
+        unsafe { __validate_gt_remainder(lhs, rhs) };
     validate_in_range::<_, MOD_BITS>(result);
-
```
Comment (kashbrti) [inline src/fns/constrained_ops.nr:336]: moved the constraining of the output of `__validate_gt_remainder` to a separate function called `check_gt_with_flags`.

File: src/fns/unconstrained_helpers.nr [hunk 2]
```diff
@@ -41,14 +41,19 @@ pub(crate) unconstrained fn __validate_in_field_compute_borrow_flags<let N: u32,
 pub(crate) unconstrained fn __validate_gt_remainder<let N: u32>(
```
Comment (kashbrti) [inline src/fns/unconstrained_helpers.nr:41]: `__validate_gt_remainder` now also returns the boolean value of the underflow.

File: src/fns/constrained_ops.nr [hunk 3]
```diff
@@ -572,3 +557,41 @@ pub(crate) fn umod<let N: u32, let MOD_BITS: u32>(
 ) -> [Field; N] {
     udiv_mod::<_, MOD_BITS>(params, numerator, divisor).1
 }
+
+// a comparison function. returns true if lhs > rhs and false otherwise
+pub(crate) fn cmp<let N: u32, let MOD_BITS: u32>(lhs: [Field; N], rhs: [Field; N]) -> bool {
```
Comment (iAmMichaelConnor) [inline src/fns/constrained_ops.nr:562]: I'm not sure what the convention should be, but it looks like `cmp` functions in the noir stdlib always return the `Ordering` type. Perhaps, therefore, this function should be renamed to something like `gt`?

File: src/tests/bignum_test.nr [hunk 4]
```diff
@@ -979,8 +980,54 @@ where
...
+    let mut b: Fq = BigNum::modulus();
+
+    a.limbs[0] -= 1;
+    b.limbs[0] -= 2;
+
+    assert(a < b);
+}
+
+#[test]
+fn test_cmp_BN_2() {
```
Comment (iAmMichaelConnor) [inline src/tests/bignum_test.nr:1013]: Perhaps it's also worth testing other common edge cases:
- 0 < 1
- 0 < -1
And if you're feeling adventurous:
- Choose lots of random bignums `r` and `s` check that if `r < s` then `r - (p-1)/2 > s`
    - The reason I suggest something like this is because these tests only test limbs[0] being different; what if the other limbs aren't equal?

File: src/bignum.nr [hunk 5]
```diff
@@ -393,6 +394,24 @@ where
     }
 }
 
+impl<let N: u32, let MOD_BITS: u32, Params> std::cmp::Ord for BigNum<N, MOD_BITS, Params>
+where
+    Params: BigNumParamsGetter<N, MOD_BITS>,
+{
+    fn cmp(self, other: Self) -> Ordering {
```
Comment (iAmMichaelConnor) [inline src/bignum.nr:401]: Should we also implement this for the `RuntimeBignum`?

**Review by iAmMichaelConnor — COMMENTED** [review]
Thanks for delving into this code! Just 3 minor comments :)

File: src/tests/bignum_test.nr (hunk 4 above)
Comment (iAmMichaelConnor) [reply src/tests/bignum_test.nr:1013]: Actually, I've written `0 < -1`... but I don't actually know what the convention should be for comparison.
I think in the Aztec codebase, we consider the "first half" of the field as "positive" numbers, and the 2nd half of the field as "negative" numbers. So we check whether things are < (p-1)/2.

Comment (iAmMichaelConnor) [reply src/tests/bignum_test.nr:1013]: For the `U256` type, regular integer comparison makes sense.
For the prime field types, I'm less sure...

File: src/fns/constrained_ops.nr [hunk 6]
```diff
@@ -28,6 +28,7 @@ global TWO_POW_120: Field = 0x1000000000000000000000000000000;
  * udiv_mod
  * udiv
  * umod
+ * cmp
```
Comment (iAmMichaelConnor) [inline src/fns/constrained_ops.nr:31]:
```suggestion
 * cmp 
 * check_gt_with_flags
```

File: src/fns/constrained_ops.nr (hunk 3 above)
Comment (kashbrti) [reply src/fns/constrained_ops.nr:562]: if you check `bignum.nr` I have an impl for the `Cmp` trait that returns an `ordering` object.

Comment (iAmMichaelConnor) [reply src/fns/constrained_ops.nr:562]: Oh, sorry, yes I saw that one. I just mean for this function in particular, perhaps its name should be changed to something that isn't `cmp`.

File: src/tests/bignum_test.nr (hunk 4 above)
Comment (kashbrti) [reply src/tests/bignum_test.nr:1013]: ah that's a good point, at the moment everything is being treated as usual integers between `0` and `modulus`. I think it'll be quite ugly to have two different impl depending on whether a number is positive or negative.

File: src/bignum.nr (hunk 5 above)
Comment (kashbrti) [reply src/bignum.nr:401]: yep will do.

File: src/tests/bignum_test.nr (hunk 4 above)
Comment (iAmMichaelConnor) [reply src/tests/bignum_test.nr:1013]: Sure, being consistent makes sense to me!

**Comment by kashbrti** [comment]
@iAmMichaelConnor this should be at a good place for a review!

File: src/tests/bignum_test.nr [hunk 7]
```diff
@@ -921,3 +921,56 @@ fn test_sub_underflow_regression() {
     let c = b - a;
     assert(c.limbs == [0, 0, 0]);
 }
+
+#[test]
+fn test_cmp_BN() {
```
Comment (TomAFrench) [inline src/tests/bignum_test.nr:926]: Can you move these all inside of a submodule grouping all of the tests related to `cmp`?

File: src/tests/bignum_test.nr [hunk 8]
```diff
@@ -921,3 +921,56 @@ fn test_sub_underflow_regression() {
...
+    let mut b: BN254_Fq = BigNum::modulus();
+
+    a.limbs[0] -= 2;
+    b.limbs[0] -= 1;
+
+    assert(a > b);
+}
+
+#[test]
+fn test_cmp_BN_fuzz(seed: [u8; 2]) {
```
Comment (TomAFrench) [inline src/tests/bignum_test.nr:970]:
```suggestion
fn test_cmp_BN_fuzz(seed: [u8; 5]) {
```

There's currently only ~65k potential inputs here.

File: src/fns/constrained_ops.nr [hunk 9]
```diff
@@ -27,6 +27,8 @@ use crate::params::BigNumParams as P;
  * udiv_mod
  * udiv
  * umod
+ * gt 
+ * check_gt_with_flags
  */
```
Comment (TomAFrench) [inline src/fns/constrained_ops.nr:32]: Let's just remove this comment. It's pretty meaningless.

File: src/bignum.nr [hunk 10]
```diff
@@ -392,6 +393,24 @@ where
...
+        if self.limbs == other.limbs {
+            Ordering::equal()
+        } else if comp_result {
+            // there was an underflow
+            Ordering::greater()
+        } else {
+            // there was no underflow
+            Ordering::less()
+        }
+    }
```
Comment (TomAFrench) [inline src/bignum.nr:411]: Can we push a `cmp` function into `constrained_ops`? We can then avoid the duplicated logic between here and `RuntimeBigNum`.

**Review by TomAFrench — APPROVED** [review]
Nits but LGTM

File: opcode_report.json
Comment (TomAFrench) [inline opcode_report.json:1]: You're committing unwanted files here. Please check what you're committing.

**Comment by kashbrti** [comment]
@TomAFrench you think this is in a good enough place to be merged?
